controllers/propagator/replicatedpolicy_controller.go

```diff
@@ -0,0 +1,445 @@
+	log.Info("Reconciling the replicated policy")
+
+	// Set the hub template watch metric after reconcile
+	defer func() {
+		hubTempWatches := r.DynamicWatcher.GetWatchCount()
+		log.V(3).Info("Setting hub template watch metric", "value", hubTempWatches)
+
+		hubTemplateActiveWatchesMetric.Set(float64(hubTempWatches))
+	}()
+
+	replicatedExists := true
+	replicatedPolicy := &policiesv1.Policy{}
+
+	if err := r.Get(ctx, request.NamespacedName, replicatedPolicy); err != nil {
+		if !k8serrors.IsNotFound(err) {
+			log.Error(err, "Failed to get the replicated policy")
+
+			return reconcile.Result{}, err
+		}
+
+		replicatedExists = false
+	}
+
+	rootName, rootNS, err := common.ParseRootPolicyLabel(request.Name)
+	if err != nil && replicatedExists {
+		if err := r.cleanUpReplicated(ctx, replicatedPolicy); err != nil {
+			if !k8serrors.IsNotFound(err) {
+				log.Error(err, "Failed to delete the invalid replicated policy, requeueing")
+
+				return reconcile.Result{}, err
+			}
+		}
+
+		log.Info("Invalid replicated policy deleted")
+
+		return reconcile.Result{}, nil
+	}
+
+	// Fetch the Root Policy instance
+	rootPolicy := &policiesv1.Policy{}
+	rootNN := types.NamespacedName{Namespace: rootNS, Name: rootName}
+
+	if err := r.Get(ctx, rootNN, rootPolicy); err != nil {
+		if k8serrors.IsNotFound(err) {
+			if replicatedExists {
+				// do not handle a replicated policy which does not belong to the current cluster
+				inClusterNS, err := common.IsInClusterNamespace(r.Client, request.Namespace)
+				if err != nil {
+					return reconcile.Result{}, err
+				}
+
+				if !inClusterNS {
+					log.Info("Found a replicated policy in non-cluster namespace, skipping it")
+
+					return reconcile.Result{}, nil
+				}
+
+				// otherwise, we need to clean it up
+				if err := r.cleanUpReplicated(ctx, replicatedPolicy); err != nil {
+					if !k8serrors.IsNotFound(err) {
+						log.Error(err, "Failed to delete the orphaned replicated policy, requeueing")
+
+						return reconcile.Result{}, err
+					}
+				}
+
+				log.Info("Orphaned replicated policy deleted")
+
+				return reconcile.Result{}, nil
+			}
+
+			log.Info("Root policy and replicated policy already missing")
+
+			return reconcile.Result{}, nil
+		}
+
+		log.Error(err, "Failed to get the root policy, requeueing")
+
+		return reconcile.Result{}, err
+	}
+
+	if rootPolicy.Spec.Disabled {
+		if replicatedExists {
+			if err := r.cleanUpReplicated(ctx, replicatedPolicy); err != nil {
+				if !k8serrors.IsNotFound(err) {
+					log.Error(err, "Failed to delete the disabled replicated policy, requeueing")
+
+					return reconcile.Result{}, err
+				}
+			}
+
+			log.Info("Disabled replicated policy deleted")
+
+			return reconcile.Result{}, nil
+		}
+
+		log.Info("Root policy is disabled, and replicated policy correctly not found.")
+
+		return reconcile.Result{}, nil
+	}
+
+	// calculate the decision for this specific cluster
+	decision, err := r.singleClusterDecision(ctx, rootPolicy, request.Namespace)
```

Comment:
Could you include this info in the `GuttedObject` sent from the root policy handler to avoid duplicating this work for each replicated policy?

You really only need to find out this info when the replicated policy was created, updated, or deleted from something other than this controller.

Reply:
On second thought, I think the queue won't deduplicate properly if you add this additional info in gutted object. I'm wondering if there's another way.

Reply:
Perhaps a Transform function could remove that additional info from the gutted object but cache it on this reconciler to be retrieved.

Reply:
The issue is that reconcile request is just the object name and namespace, so I guess you only have access to the `GuttedObject` in a mapper or predicate.

Perhaps, the root policy reconciler and this reconciler can have a shared `sync.Map` cache that contains the cluster decisions for a root policy, including if the remediation action is overridden.

Reply:
It might be possible. But having some mechanism to get the info from root controller to the replicated controller will be more complicated, any shared cache would need coordination which could mitigate the potential of the improved performance. Will reading from that cache be much better than recalculating the decision for the specific cluster? Keeping in mind that all the resources that it should be looking up to do that calculation should already be in the controller-runtime cache (likely optimized for multiple controllers reading from it at once), I don't think there would be a major difference.

I would prefer not to add the complexity. Part of the plan is to have additional "ways in" to a replicated reconciliation, for example from a PlacementDecision changing. I don't think those events will have the "full" decision, just a list of probably-affected clusters.

Reply:
Sure, that's okay. If we notice performance isn't as good as we expected, this is a place to look at for improvement.